Programmability - prep for merge

AlexRamRam opened the pull request.

metalicjames committed:
Implement ticket machine component

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement runtime locking shard component

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>
Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement directory library

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement broker library

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>
Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement agent library and specify runner interface

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>
Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement Lua execution environment and agentd component

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement EVM agent runner

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>
Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Add EVM runner to agent daemon

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement EVM load generator

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>
Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Add 3PC docker compose configurations

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>
Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Increase integration and unit test timeout in github actions

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

wadagso-gertjaap committed:
Complete Geth-compatible RPC implementation

Signed-off-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Add runner tests to unit tests binary

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

AlexRamRam committed:
Changes to compile in macos arm64

- use brew install curl (issues building tools/bench/3pc/evm/)
- evmone: remove -DBUILD_SHARED_LIBS=OFF flag from cmake
    clang 'ar' does not allow empty member list

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

metalicjames committed:
Fix crash bug where hex parsing algorithm didn't check array bounds

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Fix bug where gas refund did not account for gas price

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Fix bug where block timestamp was not a UNIX timestamp

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Return transaction receipt status

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

AlexRamRam committed:
use override keyword

to differentiate overridden hooks from evmc::Host

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
Remove postinstall step, do not copy hpp file to parent in build step

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
broker::interface::begin_return_type --> ticketnum_or_errcode_type

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
evm_runner: convert some member functions to static with explicit parameters

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
use if/else blocks in agent::impl::do_start()

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
rename agent::impl::handle_try_lock_request() --> do_try_lock_request()

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
cast to unsigned long to make clang++ happy

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
Change symbols named "dry_run" to "is_readonly_run"

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
remove accessor function evm_host::get_tx_receipt()

Dangerous as the state of the m_receipt object cannot be guaranteed.
Currently unused

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

HalosGhost commented on src/3pc/agent/runners/evm/http_server.cpp:
@AlexRamRam so, I understand that this patch removes the callback parameter from `exec_tx`, but I'm not clear on why. Is this because it was unused? It looks like an error-reporting path was taken through the callback and is no longer present. Is that readded/replaced in a different patch?

HalosGhost replied:
@metalicjames happy to have your insight here as well!

AlexRamRam replied:
Thanks @HalosGhost. Please see 
[In http_server::exec_tx() fix error propagation via json value object](https://github.com/mit-dci/opencbdc-tx/pull/213/commits/887ccbee0d7261ce9274b006e7f82f41de11fe8c)

HalosGhost replied:
This looks much better. Is there a specific motivation for `res_success_cb` to be passed as `const&`?

AlexRamRam replied:
If I recall, the clang linter inspired this change.

maurermi committed:
Hanging `dry_run` instance, update for Doxygen

Changed one remaining instance of dry_run to is_readonly_run
in src/3pc/agent/runners/lua/impl.hpp

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

maurermi committed:
Linting fixes

Additionally updates .clang-tidy to allow for final and
override together.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

AlexRamRam committed:
In http_server::exec_tx() fix error propagation via json value object

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
JSON RPC method eth_estimateGas is not supported - move to handle_static()

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

maurermi committed:
Remove parse_bytes32 (obsolete method)

Removes parse_bytes32() from src/3pc/agent/runners/evm/util.hpp
and src/3pc/agent/runners/evm/util.cpp, as this method is
effectively made obsolete by from_hex<T>(), found in
src/3pc/agent/runners/evm/util.hpp. Undoes commit
@a260bd0.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

maurermi committed:
Return errors to user in http_server

Addresses earlier comments about returning errors to user through
http_server. Returns error so the user may see what was wrong with
the submitted request.

Uses the same mechanism (JSON/callback) that is used elsewhere in
http_server for consistency.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

maurermi committed:
Address remaining review comments

General cleanup changes of agent/runners/evm/impl.cpp.
Addresses changes requested by previous review.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

maurermi commented on src/3pc/agent/runners/evm/impl.cpp:
```diff
@@ -155,10 +155,35 @@ namespace cbdc::threepc::agent::runner {
         }
         auto tn = to_uint64(maybe_tn.value());
         auto tn_key = m_host->ticket_number_key(tn);
+        auto cb
```
Note: I am not fully sure if this is what was meant by the comment to "split the nested lambdas into separate functions" -- would be good to have a second opinion

AlexRamRam replied:
Thanks @maurermi . May I suggest we replace the new `cb()` with a separate private member function as follows:

`evm_runner::lock_tx_receipt( const cbdc::buffer& valid_txid ) { ... }`

Let me now what you think.

maurermi replied:
@AlexRamRam I think that is a fair change, would help with readability

maurermi replied:
@AlexRamRam I realized that this may not be necessarily feasible because `cb` requires `tn` to be in its scope. That would indicate that if it were made a separate method, it would require tn to be passed as an argument or for `tn` to be stored at class or global scope. It cannot be passed as an argument because `m_try_lock_callback` expects a function that takes one argument of type `const broker::interface::try_lock_return_type&`. I think rather than storing `tn` in a broader scope, perhaps it's simpler to just rename `cb` to a more informative name, (ex. `handle_tx_lock_receipt_callback` ?)

AlexRamRam committed:
Add end-to-end integration test for 3PC EVM runner

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

maurermi committed:
Update evm JSON error codes to align with standard

Change the JSON-RPC error codes in the evm runner
to better align with the standards specified in
https://www.jsonrpc.org/specification#error_object

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

maurermi committed:
Add member function lock_tx_receipt to agent/runner/evm

Adds the member function lock_tx_receipt to
src/3pc/agent/runner/evm/impl.cpp. Addresses feedback
suggesting splitting of nested lambdas in run_get_block()
by creating a new member function to acquire
receipt lock, with the key goal of improving clarity.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

AlexRamRam committed:
Rename header guard prefix

Replace CBDC_UNIVERSE0 prefix with OPENCBDC_TX

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
Add more erc20 tests to 3pc integration test, fix sample C++ ABI interface

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
Confirm EVM instance is ABI compatible with the current EVMC API

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
Documentation for Programmability Architecture

- Restructure primary README
- Move Atomizer/2PC run the code section to a separate doc
- Add programmability architecture writeup and diagrams
- Add placeholder for user manual

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
Set transaction receipt status to 1 for native value transfers

Also document to clarify native value transfers in evm_host

Fix #214

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

maurermi committed:
Adds programmability user guide

Introduces a user guide for interacting with 3PC, based off a guide
written by @wadagso-gertjaap.

Update lint.sh to ignore .md files for trailing whitespace.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>

maurermi committed:
Adds 3pc demo tools, and a script to run locally

Adds a sample ERC20 contract and deploy script to
go along with the programmability user guide. Adds a
script to run 3PC locally, and updates the 3pc
docker-compose file to reflect OpenCBDC image naming
convention.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

AlexRamRam committed:
Gracefully return error when write-locks requested for read-only transactions

Also:
- Rename agent state code `function_error` -> `function_exception` to
  distinguish from `function_failed`. `function_exception` conveys an
  unexpected system error
- `agent::impl::do_try_lock_request()` & `broker::impl::try_lock()` - clarify
  that false return value is returned only upon unexpected exceptions

Fix #216

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

HalosGhost commented on README.md:
```diff
@@ -18,8 +18,9 @@ The design decisions we made to achieve these goals will help inform policy make
-
-We explored two system architectures for transaction settlement, both based on an [unspent transaction output (UTXO)](https://en.wikipedia.org/wiki/Unspent_transaction_output) data model and transaction format.
+There are two main architectures as follows:
+## UTXO & UHS-Based Transaction Processor
```
```suggestion
We have explored several architectures as follows:

## UHS-Based
```

HalosGhost commented on README.md:
```diff
@@ -36,8 +37,19 @@ Both architectures handle multiple geo-distributed datacenter outages with a [re
     - Geo-replicated latency <1 second.
 
-Read the [architecture guide](docs/architecture.md) for a detailed description of the system components and implementation of each architecture.
+Read the [2PC & Atomizer architecture guide](docs/architecture.md) for a detailed description of the system components and implementation of each architecture.
```
```suggestion
Read the [UHS architecture guide](docs/architecture.md) for a detailed description of the system components and implementation of each architecture.
```

*Note*: this is probably a good time to move `docs/architecture.md` to something like `docs/uhs-architecture.md` to be less confusing.

HalosGhost commented on README.md:
```diff
@@ -36,8 +37,19 @@ Both architectures handle multiple geo-distributed datacenter outages with a [re
-Read the [architecture guide](docs/architecture.md) for a detailed description of the system components and implementation of each architecture.
+Read the [2PC & Atomizer architecture guide](docs/architecture.md) for a detailed description of the system components and implementation of each architecture.
+
+## Generic Smart Contract Platform ("3PC")
```
Here and, in all other docs, please keep to the same style used in the rest of the document (e.g., unify the numbering/bulleting style with the section above), and ensure there is only one logical sentence/thought-per-line. In particular, if there's a `.`, you should probably add a new-line after it.

HalosGhost commented on README.md:
````diff
@@ -117,16 +130,16 @@ To run the system with our pre-built image proceed to the [next section](#launch
         $ docker compose --file docker-compose-2pc.yml up --build
         ```
-    1. Atomizer architecture:
+    2. Atomizer architecture:
````
This is not necessary; markdown almost always automatically numbers lists; it's a *really* good idea to get in the habit of numbering all lists as `1.` so that you don't need to make changes like this. (This is what is done in lists above.)

HalosGhost commented on README.md:
````diff
@@ -73,8 +85,9 @@ Note that if you have not already installed the xcode cli tools you will need to
 ```
-
-# Run the Code
+# Run the Code: Programmability architecture
````
This is jarring. Either the basics of getting started with 3PC should belong in the same place as the UHS architectures here in the readme, or the UHS architecture getting-started routine should also be separated to mirror the programmability stuff.

(non-blocking)

HalosGhost commented on benchmarks/transactions.cpp:
```diff
@@ -97,10 +97,14 @@ static void Nto1_tx(benchmark::State& state) {
-        ASSERT_EQ(wallet_b.count(), 1);
+        ASSERT_EQ(
+            wallet_a.balance(),
+            static_cast<unsigned long>(SWEEP_MAX * 2 - 2 * state.range(0)));
```
Am I right in thinking that all these with an explicit cast were not fixable by adding the literal suffixes? If not, feel free to immediately mark this as resolved.

If, however, these can be silenced with the suffix addition, that's definitely a little less noisy.

AlexRamRam replied:
Cannot use UL suffix here for the 2nd value because it's not a literal.

HalosGhost commented on docs/programmability_architecture.md:
Reformat. In particular, follow standard document structure (use progressive headings, rather than skipping ones), and keep to a single-line per-sentence.

HalosGhost commented on docs/programmability_architecture_diagram.jpg:
Here, and below, do not check-in binaries, only source-documents. If source-documents are not available or are in proprietary formats, they need to be rebuilt. Git is not for binaries.

AlexRamRam replied:
Git is not for derived elements - agreed.
In this case - due to time and other constraints - the architecture diagram was generated using a vendor software program. If the value of this diagram does not adequately justify its existence as is, we can remove it. Or, we can accept it as the source image at least temporarily until we redraw using another tool.

AlexRamRam replied:
Investigating options:
* use mermaid & commit source
* commit svg output created by [inkscape](https://inkscape.org/)

AlexRamRam replied:
As an imperfect but acceptable solution, replaced .jpg binaries with formatted .svg files which can be loaded into an open source design tool such as [Inkscape](inkscape.org) and be modified.

HalosGhost commented on docs/programmability_architecture.md:
```diff
@@ -0,0 +1,33 @@
+
+The lower-level layer is a generic distributed key-value data store providing [ACID](https://en.wikipedia.org/wiki/ACID) database transactional properties. This data store is generic as it does not impose constraints on the type or the size of the data it handles, and is agnostic to its client execution layer. The end user does not directly interact with this layer. The implementation consists of ...
+
+The higher level computation layer executes programs (i.e. smart contracts) and uses the lower-level key-value data store to maintain its state. The _Agent_ is the main interface for this layer and serves the role of a transaction coordinator. This computation layer contains implementations of the _Runner_ interface which is an abstraction for retrieving and executing context-specific bytecode. T ...
```
```suggestion
The higher level computation layer executes programs (i.e. smart contracts) and uses the lower-level key-value data store to maintain its state. The _Agent_ is the main interface for this layer and serves the role of a transaction coordinator. This computation layer contains implementations of the _Runner_ interface which is an abstraction for retrieving and executing context-specific bytecode. The system provides two _Runner_ implementations ([EVM](https://ethereum.org/en/developers/docs/evm/) and [Lua](https://www.lua.org/) and demonstrates the flexibility of the architecture to support different program execution environments.
```

Note the removal of the extra paren (making such diffs more clear is the specific reason why it is good practice to keep to a single sentence-per-line).

AlexRamRam replied:
Paren is necessary - but it's currently part of Lua hyperlink (asymmetric). Subtle - hard to see in the viewable doc. Good catch.
Excellent point on keeping single sentence per line.

HalosGhost commented on docs/programmability_architecture.md:
```diff
@@ -0,0 +1,33 @@
+### JSON-RPC Server
+The _JSON-RPC Server_ implements the [Ethereum JSON-RPC API](https://ethereum.org/en/developers/docs/apis/json-rpc/) over HTTP. This is the external interface compatible for external Ethereum clients such as wallets (e.g. Metamask), and development tools (e.g. Hardat, Truffle, ethers.js, web3.js)
+### Runner
+The _Runner_ is an abstraction for retrieving and executing context-specific bytecode. The system provides two _Runner_ implementations ([EVM](https://ethereum.org/en/developers/docs/evm/) and [Lua)](https://www.lua.org/) and demonstrates the flexibility of the architecture to support different program execution environments.
```
```suggestion
The _Runner_ is an abstraction for retrieving and executing context-specific bytecode. The system provides two _Runner_ implementations ([EVM](https://ethereum.org/en/developers/docs/evm/) and [Lua](https://www.lua.org/) and demonstrates the flexibility of the architecture to support different program execution environments.
```

Another additional parenthesis. Maybe it's worth re-reading the docs for copy-editing?

AlexRamRam replied:
Paren is necessary - but it's currently part of Lua hyperlink (asymmetric). Subtle - hard to see in the viewable doc. Good catch.

HalosGhost commented on docs/programmability_architecture.md:
```diff
@@ -0,0 +1,33 @@
+### Directory
+The _Directory_ identifies the target _Shard_ unit for each and every key-value data pair. This is a detail managed by the _Broker_, allowing the end-client (the _Agent_) to interact with the distributed _Shards_ as a single database unit. The _Directory_ is where the sharding logic is defined and its implementation can be designed to optimize database resources for particular use cases.
+### JSON-RPC Server
+The _JSON-RPC Server_ implements the [Ethereum JSON-RPC API](https://ethereum.org/en/developers/docs/apis/json-rpc/) over HTTP. This is the external interface compatible for external Ethereum clients such as wallets (e.g. Metamask), and development tools (e.g. Hardat, Truffle, ethers.js, web3.js)
```
> Hardat

Hardhat typo?

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,231 @@
+
+Make sure you have [Docker installed](https://docs.docker.com/engine/install/ubuntu/).
+
+Make sure you have [NodeJS v14+ installed](https://github.com/nodesource/distributions/blob/master/README.md).
```
I got a warning for using Node v19; it all seemed to work fine, but if we have specific version requirements, we should specify them as clearly as we can.

In the ideal case, we'd include a `.nvmrc` (cf. https://github.com/mit-dci/opencbdc-ui/commit/e9bbd8f0d12de42bd04c829bbf950949f07057c5).

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,231 @@
+
+## Clone the repository
+
+First, clone OpenCBDC locally, and pull the programmability branch:
```
This is repeated documentation from the README, I think it makes more sense to have the README include cloning directions, and leave it out of this guide. Additionally, note that this is written with the understanding that the code sits at the `programmability` branch, which won't be the case when this lands. Please rephrase so that everything references the correct locations for where everything will live *post-merge*.

HalosGhost commented on docs/programmability_user_guide.md:
````diff
@@ -0,0 +1,231 @@
+
+```
+docker build --target base . -t ghcr.io/mit-dci/opencbdc-tx-base:latest
+```
````
````suggestion
```console
$ docker build --target base . -t ghcr.io/mit-dci/opencbdc-tx-base:latest
```
````

Where possible, use appropriate syntax highlighting, and make clear when commands are to be run at the terminal (by convention, this is done by showing the `$` as the prompt for non-root commands (some flavorful folks who prefer zsh will use `%` instead here), and `#` for root-required commands).

HalosGhost commented on docs/programmability_user_guide.md:
````diff
@@ -0,0 +1,231 @@
+### Start it up
+
+```
+# will build containers if not already built
````
If this builds the images if they don't already exist in all cases, then do we need the build step? If it doesn't build in all cases, should we have the comment?

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,231 @@
+
+Select to create a javascript project.
+
+If you are using Docker, edit your hardhat.config.js to look like this:
```
Can we ship these two config files and replace this documentation with a copy command? If not, why not? Is there more explanation needed here?

HalosGhost commented on docs/programmability_user_guide.md:
````diff
@@ -0,0 +1,231 @@
+```
+* Note: the value in the `accounts` array is the private key of one of the hard-coded, pre-minted accounts in 3PC/EVM.
+
+You should now be able to use hardhat through for instance
````
> for

?

HalosGhost commented on docs/programmability_user_guide.md:
````diff
@@ -0,0 +1,231 @@
+npx hardhat console
+```
+
+Look at more fun stuff with hardhat and compiling and deploying contracts [here](https://hardhat.org/hardhat-runner/docs/getting-started#quick-start)
````
In general, avoid using `here` or similar things as link-text (terrible for accessibility), consider rephrasing this to be something like "See [the upstream documentation](thelink) for more fun..."

HalosGhost commented on docs/programmability_user_guide.md:
````diff
@@ -0,0 +1,231 @@
+}
+```
+
+Note that the `>` symbol is the hardhat prompt. The two commands to issue are `const signers = await ethers.getSigners()` and then `await signers[0].sendTransaction(...)`
````
This probably should preface the codeblock in-question so it serves as instruction; that'd also serve as an opportunity to remove the repetition without any loss of clarity.

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,231 @@
+
+Modify the provided contracts/MITCoin.sol file to deploy your own ERC20 token using the below directions, or create one [from scratch](https://dev.to/yakult/a-concise-hardhat-tutorial-part-2-writing-erc20-2jpm).
+
+First, update the provided deploy script (`scripts/deploy.js`) to reflect the new ERC20 token.
```
This is probably clear for people who are quite familiar with solidity, but probably isn't enough information otherwise; either, we should link to upstream documentation strategically to onboard people as-necessary, or we should be far more specific.

Note also that, in following these instructions, I received a `contracts` and `scripts` directory underneath the playground, but these instructions seem to refer to those directories as they exist in the root of the repo; either way is fine, but we should ensure the right files exist in the right places for all the commands to work.

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,231 @@
+npx hardhat console
+> const yourToken = await (await ethers.getContractFactory("*yourTokenName*")).attach("*yourTokenAddress noted above*");
+undefined
+> await cbdcToken.mint("*your metamask address*", ethers.utils.parseUnits("5000"))
```
```suggestion
> const yourToken = await (await ethers.getContractFactory("*yourTokenName*")).attach("*yourTokenAddress noted above*");
undefined
> await yourToken.mint("*your metamask address*", ethers.utils.parseUnits("5000"))
```

just renamed the variable in the second line to match the declaration from the first.

HalosGhost commented on docs/programmability_user_guide.md:
````diff
@@ -0,0 +1,231 @@
+
+```
+npx hardhat console
+> const yourToken = await (await ethers.getContractFactory("*yourTokenName*")).attach("*yourTokenAddress noted above*");
````
Note that "token address" doesn't appear anywhere in the output from the commands. Let's make sure the specific properties are referred to correctly (ideally with example output, but as long as the names are correct, examples are a stretch goal).

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,231 @@
+flag in the agent0 command argument, in the file `docker-compose-3pc.yml`.
+
+### Usage pre-merge (current state)
+The Dockerfile uses a remote image as the base image (`ghcr.io/mit-dci/opencbdc-tx-base:latest`). This will be updated when this branch is merged into trunk.
```
In the future, this should be avoided. Reviewers should be reviewing documentation as it will exist in the repo; exceptional instructions, if they need to exist for testing, should be provided in the PR body/comments.

HalosGhost commented on scripts/3pc-run-local.sh:
```diff
@@ -0,0 +1,47 @@
+LOGLEVEL="WARN"
+
+function print_help() {
+    echo "Usage: myscript.sh [OPTIONS]"
```
> myscript.sh

HalosGhost commented on scripts/lint.sh:
```diff
@@ -4,7 +4,7 @@ set -e
 
 check_files=$(git ls-files \
-              | grep -v -E ".jpg|3rdparty" | cat)
+              | grep -v -E ".jpg|.pdf|3rdparty" | cat)
```
This should be reverted at the same time as all binaries are removed.

AlexRamRam replied:
pdf is not necessary and will be removed. Discussions about new .jpg diagrams are elsewhere.

HalosGhost commented on scripts/wait-for-it.sh:
```diff
@@ -0,0 +1,182 @@
+WAITFORIT_CHILD=${WAITFORIT_CHILD:-0}
+WAITFORIT_QUIET=${WAITFORIT_QUIET:-0}
+
+# Check to see if timeout is from busybox?
+WAITFORIT_TIMEOUT_PATH=$(type -p timeout)
```
Do we want this variable definition that gets immediately overwritten?

AlexRamRam replied:
It's used as input in the next line - can be prettier, but ok with me.

HalosGhost replied:
Definitely a misread from me. Ignore this noise.

HalosGhost commented on src/3pc/agent/impl.cpp:
```diff
@@ -0,0 +1,679 @@
+            m_ticket_number.value(),
+            payload,
+            [this](broker::interface::commit_return_type commit_res) {
+                if(commit_res.has_value()) {
+                    if(std::holds_alternative<
+                           runtime_locking_shard::shard_error>(
+                           commit_res.value())) {
+                        auto err
+                            = std::get<runtime_locking_shard::shard_error>(
+                                commit_res.value());
+                    }
+                }
```
Is this block necessary? It seems like `err` is unused, and getting that value seems to be the only point of this block.

AlexRamRam replied:
Not needed - will be removed. Handled in `handle_commit()`

HalosGhost commented on src/3pc/agent/impl.cpp:
```diff
@@ -0,0 +1,679 @@
+                    m_log->fatal("Result reported in rollback_complete state "
+                                 "when result is not retry");
+                }
+                [[fallthrough]];
```
Since these cases should all represent fatal bugs, is there a reason we fall-through to `break`, seemingly to retry, instead of exiting?

Additionally, since all these cases fall-through, would hitting one of the earlier cases result in all these error messages being logged?

AlexRamRam replied:
Note that `m_log->fatal(...)` calls system `exit()` (i.e. execution does not return to this stack, there is no retry).

The log object having the power to system exit is surprising - something we can discuss and potentially change. e.g. rename `fatal()` to `fatal_system_exit()`

It's curious that we're not getting the `fallthrough` warnings, some cleanup is needed here.

HalosGhost replied:
Funny that you mention it is surprising `logger::fatal()` is able to fail. That surprises me right now, as a matter of fact. That's also on me for not having investigated whether or not it would terminate; clearly came from an untested response. Consider adding a closure task for this review item to open a bug for potentially reconsidering fatal-logging's immediate termination. Maybe it's not large enough a deal to worry about, but I'd like to not forget about it and avoid revisiting due to attrition.

AlexRamRam replied:
Will create an issue item for this and close.

AlexRamRam replied:
#219

HalosGhost commented on src/3pc/agent/impl.cpp:
Not actually a comment on specifically this file, though it stuck out particularly as I was reviewing this one: `[[fallthrough]]` should *always* [be explicit](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-break).

AlexRamRam replied:
... I can think of cases where I would prefer to not see `[[fallthrough]]`

HalosGhost replied:
Not currently a hard-and-fast rule, we generally try to adhere to the [CppCoreGuidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). That generally governs our choices for linting options, as it helps us define a reasonable code quality and formatting standard.

To butcher a saying from the Go programming community, "The CppCoreGuidelines are no one's favorite formatting/practice guide; but having a standard practice guide to follow is everyone's favorite." Consider this case to be closer to a nit than not.

AlexRamRam replied:
Create issue item for this & close

AlexRamRam replied:
#220

AlexRamRam replied:
Following is the case I had in mind where I would prefer to not see `[[fallthrough]]`. [CppCoreGuidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) is OK with this and `clang-tidy` requires them to be removed (complains when `switch has 2 consecutive identical branches`):

```
Multiple case labels of a single statement is OK:

switch (x) {
case 'a':
case 'b':
case 'f':
    do_something(x);
    break;
}
```

HalosGhost commented on src/3pc/agent/runners/evm/host.hpp:
```diff
@@ -0,0 +1,210 @@
+                 bool is_readonly_run,
+                 interface::ticket_number_type ticket_number);
+
+        [[nodiscard]] auto
```
Here, below, and elsewhere, public methods should all have doc comments.

(non-blocking, but strongly encouraged)

maurermi replied:
The non-documented public methods in host.hpp are all inherited methods from the class evmc::HostInterface in evmc.hpp, which then refers using @copydoc to the associated functions in evmc.h 

Would you recommend that in this case, we include new doc comments within the OpenCBDC codebase for these inherited functions? Curious what the protocol is here

HalosGhost replied:
Discussed offline. For any method including `override`, it is assumed the core documentation is elsewhere, and only relevant deviations from the overridden method are documented inline.

In this case, these methods are overriding for the purpose of providing our own implementation of an external interface, so the deviations are all implementation-details. To that end, it's fair for these to remain undocumented here.

@AlexRamRam and @maurermi, feel free to chime in here with any thoughts or clarifications if I missed any context which should be here for traceability.

HalosGhost commented on src/3pc/agent/runners/evm/init_addresses.hpp:
Is there any reasoning/clarification/documentation for why these addresses rather than others, or why any addresses need to have accounts minted for them in-advance?

I see that it's used by `mint_initial_accounts()`; is this for ease-of-testing/dive-in?

AlexRamRam replied:
It's convenient and useful to have fully deterministic execution for testing. I'm ok with this.

HalosGhost replied:
I have no issue with using these for testing. If this is exclusively for testing, is there a reason it must be in `src`? Either way, can a doc comment be added explaining what these are for? (That's the easiest path to close this review task in my opinion.)

AlexRamRam replied:
Will document the intent of this and close

HalosGhost commented on src/3pc/agent/runners/evm/serialization.hpp:
```diff
@@ -0,0 +1,171 @@
+    /// \param logger logger to output any parsing errors to
+    /// \param chain_id the expected chain ID for the transaction. If the
+    // transaction contains a different chain ID this method will return
+    // std::nullopt
```
Here, and below, there are broken doc comment line continuations.

HalosGhost commented on src/3pc/agent/runners/interface.hpp:
```diff
@@ -0,0 +1,169 @@
+        /// Begins function execution. Retrieves the function bytecode using a
+        /// read lock and executes it with the given parameter.
+        /// \return true.
+        virtual auto run() -> bool = 0;
```
Is there a reason to have this return a value at all? The core interface, and both runner implementations define it to always return `true`.

AlexRamRam replied:
This is an appropriate interface to return a system level failure (although not used). However, should be `[[nodiscard]]` and aided by comments.

HalosGhost commented on src/3pc/directory/impl.hpp:
```diff
@@ -0,0 +1,35 @@
+#include "interface.hpp"
+
+namespace cbdc::threepc::directory {
+    /// Implementation of a directory. Uses siphash to uniformly map keys to
```
Nit: this will only be uniformly-distributed if `m_n_shards` is a power-of-two; else, there will be bias introduced from `%`. In this case, I assume this is likely immaterial since it seems unlikely that anything in the system depends on key distribution being uniform so much as being generally load-balanced across the shards.

AlexRamRam replied:
`... uniformly map keys to shard IDs based on a fixed number of shards`: seems correct to me (perhaps out of ignorance). @HalosGhost: if a value-adding correction is needed please make the change, or close.

HalosGhost replied:
Sorry, allow me to clarify: [`impl.cpp`:15](https://github.com/mit-dci/opencbdc-tx/pull/213/files#diff-77a7dfffad078695ca8ec3a68263414d82b8790b84ffdc33b307689e2b69eba9R15) divides the key generated by the number of shards running to select the shard for load-balancing. This constrains the random-number-generator of the hash function to the range of running shard ids (i.e., `[0, m_n_shards - 1]`). However, it introduces "modulo bias".[^1]

[^1]: https://stackoverflow.com/questions/10984974/why-do-people-say-there-is-modulo-bias-when-using-a-random-number-generator

I mentioned this because the comment is inaccurate. However, as I mentioned, I do not know if any security properties of the system depend on this property (i.e., I am not sure if this load-balancing is required to be perfectly uniformly distributed). If it does not matter, and this is either acceptably uniform for the purposes of the architecture, or the uniformity of the distribution is only desirable, not necessarily something that must be achieved, then that's not a problem! The comment should be corrected though.

However, if the load-balancing here is needed to be uniformly random, I see two valid directions: 1) we should document that the number of shards is required to be a multiple of the range of the hash-function. This is equivalent to taking the `n`-least-significant bits as the shard index. Or, 2) we modify `key_location()` cut the bias from this selection. This could be done either with the STL's templates for random-number generation (which we use extensively elsewhere in this codebase, for easy reference; [Edit]: or maybe one of our secure random sources would be even better), or, say with rejection sampling into a large range known to be a multiple of `m_n_shards`, and then taking the modulus just as it does now. That also should perfectly remove the bias since it creates an equivalent case to `m_n_shards` being a multiple of the hash-function's range.

If we don't know, or don't need to know at this level of depth, whether or not this property is important to the system's security properties, we could potentially close this task by modifying the comment to be accurate. That's not a hard fix or unreasonable, but I don't know if it's the right choice. Definitely happy for alternative actions too, of course.

HalosGhost commented on src/3pc/runtime_locking_shard/state_machine.cpp:
```diff
@@ -0,0 +1,99 @@
+        -> nuraft::ptr<nuraft::buffer> {
+        auto maybe_req = from_buffer<rpc::replicated_request>(data);
+        if(!maybe_req.has_value()) {
+            // TODO: This would only happen if there was a deserialization
```
Have the `TODO`s sprinkled throughout the code been evaluated for whether or not they should be made into issues or addressed now rather than later?

AlexRamRam replied:
Good point that some of the TODO's can be made more transparent as GH issues, though certainly not all. What has been the policy for prior commits?

HalosGhost replied:
This is also something we don't have a hard and fast rule for. I do want to, however, try to help give people easier ways to dive into the code, as well as to unify on as close to a single source of truth for people to discover a comprehensive list of known improvements available to be made to the code base.

I recognize that some things are going to be worthy of being added as items to #3, and nothing more, some probably aren't worth any additional documentation at all (in which case, maybe we should consider removing the TODO?).

This is both a question of "What is the list of TODOs which we think should be turned into issues to increase their accessibility?" (Which includes the possibility "there aren't any worth bringing over.") and a request to improve the documentation for those that are worth keeping.

The invovled part is actually determining which of the newly-added TODO lines are worth trying to document more sustainably/visibly. Once that's been determined, and recorded, even adding those items to an omnibus issue (similar to #3) would be a reasonably closure path.

hgyami replied:
@AlexRamRam @maurermi can you put together a comprehensive list of the TODOs in one place that we can review in the eng meeting tomorrow? Will also raise with Bob - will be helpful as we build our research agenda/roadmap moving forward!

HalosGhost commented on src/3pc/ticket_machine/interface.hpp:
```diff
@@ -0,0 +1,63 @@
+
+    /// \brief Ticket machine interface.
+    ///
+    /// Returns batches of monotonically increasing ticket numbers for
```
Nit: this is not monotonically increasing; it's expected to wrap (just not frequently).

AlexRamRam replied:
I believe the existing comment is appropriate for a general purpose statement for the interface. Note that this detail (wrapping, etc) are documented in `get_ticket_number()`

HalosGhost commented on src/3pc/util.cpp:
```diff
@@ -0,0 +1,247 @@
+        return ret;
+    }
+
+    auto parse_args(int argc, char** argv)
```
This is a little error-prone and definitely non-standard. We need to reunify configuration between the new architecture and the old infrastructure; when we do, we should also settle on a standardized argument/option parser.

(non-blocking)

AlexRamRam replied:
Agreed. Also, keeping in mind that corresponding changes to test controller are likely required.

HalosGhost replied:
I mentioned this as an issue without offering a clear action to close review item; sorry about that!

In short, I mentioned this because I think the addition of this architecture exacerbates the existing DIY configuration system, which motivates our trying to solve the problem. I think a new issue is warranted, targeting a unifying configuration system for the three architectures we have, and with space for more. (Probably referencing #173 and/or #81.) At some point, before merge, that issue should be created (or establishing an alternative path toward handling this inconsistency; for example, an official sunset, or stable release plan---these are not the only options, of course).

Hopefully that clarifies the purpose of this task, as well as a clear completion target.

AlexRamRam replied:
#218

HalosGhost commented on src/util/rpc/http/kqueue_event_handler.cpp:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    auto kqueue_event_handler::init() -> bool {
+        if((m_kq = kqueue()) == -1) {
+            return false;
+        }
+        return true;
```
```suggestion
        m_kq = kqueue();
        return m_kq != -1;
```

Style nit; no reason to prefer inline-assignment inside the `if`, especially when the `if` is of the always-redundant form `if (condition) { return bool; } else { return bool; }`

AlexRamRam replied:
Agreed on separating `m_kq = kqueue();`. However, I suggest we keep the if/else structure for clarity. At least for me reading `return m_kq != -1` (unlike, let's say `return x==y`), despite its brevity, requires some brain cycles.

HalosGhost replied:
(Sorry if this seemed silly; I realize it probably did. It's just one of those ones that bothers me for some reason. Take nothing personal about it, please!)

HalosGhost replied:
It bothers me, but as it is marked as a nit, you can safely read it as "I'm not particularly moved by this one; take it or leave it"; so, if you don't find this compelling, please do feel close it and I won't be mad in any fashion.

HalosGhost commented on src/util/rpc/http/kqueue_event_handler.cpp:
```diff
@@ -0,0 +1,113 @@
+                struct kevent ev {};
+                EV_SET(&ev, fd, EVFILT_READ, EV_DELETE, 0, 0, 0);
+                m_evs.emplace_back(std::move(ev));
+            }
+                {
```
If `clang-format` is responsible for this really odd formatting, we should figure out how to fix it.

(long-term todo/issue; non-blocking)

AlexRamRam replied:
will first run clang-format on this and see if it fixes it. otherwise will create an issue item and close

AlexRamRam replied:
#221

HalosGhost commented on tools/bench/3pc/lua/lua_bench.cpp:
```diff
@@ -0,0 +1,350 @@
+    auto pay_contract = cbdc::buffer();
+    pay_contract
+        = cbdc::buffer::from_hex(
+              "1b4c7561540019930d0a1a0a040808785600000000000000000000002877400"
```
Here, and elsewhere, can we avoid using particularly large magic hex strings? Ideally, we'd have the plaintext version in source-control, and compile/convert/serialize them appropriately at init.

maurermi replied:
Currently working on a fix for this

maurermi replied:
As a follow up, is there any reason that comes to mind for store the bytecode in plaintext anywhere? It's fairly easy to generate the bytecode at runtime by leveraging Lua's embedded scripting capabilities and then we don't need to store this massive string anywhere persistently. 

So far I have two solutions: one where this bytecode string is generated during the build process and is stored persistently in a text file, which is then read in at runtime -- and another where the bytecode is generated at runtime (not requiring any external storage, or large static strings). I prefer the second solution but if there's any reason to store the bytecode persistently, I would be interested to hear it.

HalosGhost replied:
Unless any of the generation would happen during performance measurements, I see no reason to keep around the hex-version if it's just as easy to keep around the source.

AlexRamRam replied:
Unless there is a reason to re-compile every time, compile once and keep the derived (non-git element) persistent file.

maurermi replied:
@AlexRamRam Sorry just now seeing this -- I would say in regards to the bench program it makes sense to interpret the Lua script at runtime because it allows easier interaction with different Lua scripts which could be useful, but in terms of the unit test set up I can see interpreting the script at runtime being unnecessary

maurermi replied:
^^ Also -- not sure if I did a good job explaining this part, but the bytecode is only generated once at runtime (in the setup for either the bench test or unit test) and then it exists statically for the remainder of the test, it is not being regenerated each time the bytecode is used

HalosGhost commented on tools/bench/3pc/evm/contracts.hpp:
This file, because it is generated, is partly exempt from the hope to have all public methods documented. However, given that it is generated, should we actually check it in at all, or should we add a build-step so that this file is *always* generated as-needed?

(non-blocking)

AlexRamRam replied:
Fully agree in principle (to have user build/generate this). This would add more dependencies to the build (node, hardhat, etc), but cleaner and worth considering.

HalosGhost replied:
Mmm, very good point. I don't love adding new dependencies (and this PR includes quite a few additional build-time ones already). I suppose I hadn't considered they had been excluded for that reason because we already pull in several EVM-specific dependencies at configure-time. It's for this same reason I hadn't really considered there to be a separate lifecycle phase between building the full repository at the top-level and doing anything lower down.

Do you think we should make that explicit? If so, would the reasonable separation be between the C++ architecture at build-time, and then runner-specific build-time should happen later and on-demand? This would allow people who are interested only in, say, the EVM to avoid downloading lua-dependencies if they have no intention of using those (or, of course, vice versa). This would require a bit of build-system work (e.g., some `./configure.sh`-style infrastructure for each runner).

Alternatively, we could life all the dependencies from all committed runners to top-level, and unify the build-times. This is *much* easier, both to write, and to maintain. The downside is that, the officially-supported build infrastructure will assume all users will download all dependencies of all committed code. And, to be fair, that might not be a sustainable long-term view. (I know it doesn't immediately feel scalable or particularly frugal.) However, because not a lot of work has been done yet on planning for the long-term maintainability of multiple, potentially wildly-disparate architectures, I do not think this particular case needs to be solved right now, and I would definitely accept such a simplification as a valid solution for now.

HalosGhost replied:
Or, is the separation as simple as "all the C++ stuff" building at top-level and "all the other stuff" separate? I *think* that accurately captures the current separation. If so, and that's the intention, then the path to close this review task is far simpler: just make it explicit. Namely, add a step to the README's walk-through for the additional steps (or, if we separate the build guides out entirely, mention that sub-directories may have a required build-step). (This is the documentation task that, if one of the other splits makes sense to pursue, might still be required to complete.)

AlexRamRam replied:
Currently the "other stuff" (i.e. creating the contract ABI for C++) has to be done first to build the "C++ stuff" like `evm_bench` and the 3pc integration test. Instead of performing that step we have committed the derived c++ element. That part is well documented and can easily be regenerated.

maurermi replied:
Have we decided how to proceed in regard to this issue (or whether we wish to at all)?

HalosGhost replied:
I think making the current “C++ gets built at top-level, with all deps as-needed for that goal / subdirs might need other deps” explicit is the easiest, acceptable solution (likely taking the form of an **N.B.,**-style comment (which we have in several places in our docs already) in the walkthrough mentioning that modifying the system may require other dependencies, documented where-relevant).

However, I think this separation is far from ideal; personally, the following three splits are the best ergonomically:

* no separation; all deps to run/modify all code are handled by `./scripts/configure.sh` (con: wasteful; pro: single step for everyone)
* separate at the runner-level (con: big lift, requires splitting tests/deps-handling, etc.; pro: super-clear why the split is where it is, and people can install only the things they care to use
* do both! Modularize each piece so that `./scripts/configure.sh` (or, more likely, its replacement) has feature-flags which determine what gets brought in and built (this is actually what `configure.sh` does more normally) (con: massive reworking of the build-system; pro: it's much closer to the standard build experience, and also reaps all the pros of both other separations)

I think, especially given our hope to merge this ASAP, the first solution (which is just adding a single line in the docs) is probably most reasonable. I would ask that an additional closure task be added for this thread: namely, opening an issue to revisit the build system for one of these three alternative separations.

HalosGhost reviewed (changes requested):
Several parts of the documentation need a bit of attention before merge; however, the vast majority of the content looks solid.

I'll also add a reminder note that there are commits which must be squashed before merge as well.

AlexRamRam committed:
Address several style/documentation/aesthetic issues

Cleanup re [[fallthrough]] in switch statements (Fix #220)
  - src/3pc/agent/impl.cpp
  - src/3pc/broker/impl.cpp
  - src/uhs/twophase/coordinator/format.cpp
src/3pc/agent/impl.cpp
  - do_commit(): remove unused code block in lambda callback passed to m_broker->commit()
src/3pc/agent/runners/interface.hpp
& src/3pc/agent/runners/lua/impl.hpp
& src/3pc/agent/runners/evm/impl.hpp
  - runner::interface::run(): add [[nodiscard]] and comments
src/3pc/agent/runners/evm/init_addresses.hpp
  - rename init_addresses → init_addresses_for_testing
src/3pc/agent/runners/evm/serialization.hpp
  - fix broken doc comments
src/3pc/directory/impl.cpp
& src/3pc/directory/impl.hpp:
  - key_location(): note that using modulo operation creates a (very) small bias from a uniform distribution
src/util/rpc/http/kqueue_event_handler.cpp
  - init(): fix if statement containing assignment
tests/unit/3pc/agent/runners/evm/evm_test.cpp
  - Direct reader to origins of bytecode

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

temp

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

maurermi committed:
Generate lua_bench bytecode at runtime

Removes the large bytecode string for the Lua pay contract
and generates bytecode at runtime. Now expects that the
.lua file containing the contract code is provided as a
config input in lua_bench.

Makes similar change in account_test.cpp to generate
bytecode at runtime. Moves location of and slightly
modifies gen_bytecode.lua to better integrate within
c++ source code.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

maurermi committed:
Add doc comments for public methods

Adds doc comments to evm_bench and tools/bench/3pc/evm/rpc_client.
Also addresses lack of doc comments in host.hpp by pointing upstream.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,165 @@
+## Prerequisites
+
+Make sure you have [NodeJS installed](https://github.com/nodesource/distributions/blob/master/README.md).
+We reccomend using `v18` or `v20`, as Hardhat [actively supports these versions](https://github.com/nodejs/release#release-schedule) currently.
```
```suggestion
We recommend using `v18` or `v20`, as Hardhat [actively supports these versions](https://github.com/nodejs/release#release-schedule) currently.
```
typo nit.

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,165 @@
+Using Docker this will be `http://localhost:8080/`. If running outside of
+Docker, use the IP and port specified
+ when running `./scripts/3pc-run-local.sh`
+(default is `http://localhost:8888/`).
```
```suggestion
Using Docker this will be `http://localhost:8080/`.
If running outside of Docker, use the IP and port specified when running `./scripts/3pc-run-local.sh` (default is `http://localhost:8888/`).
```
nit to fix some really odd line-breaking.

HalosGhost commented on docs/programmability_architecture_diagram.svg:
Thank you for the switch to SVG, this is dramatically better! At some point, I wouldn't mind revisiting our documentation image artifacts to go to a more unified source form we can leverage universally, but this is a fabulous step towards that. :+1:

HalosGhost commented on docs/programmability_architecture.md:
```diff
@@ -0,0 +1,54 @@
+This is the external interface compatible for external Ethereum clients such as wallets (e.g. Metamask), and development tools (e.g. Hardhat, Truffle, ethers.js, web3.js)
+### Runner
+The _Runner_ is an abstraction for retrieving and executing context-specific bytecode.
+The system provides two _Runner_ implementations ([EVM](https://ethereum.org/en/developers/docs/evm/) and [Lua)](https://www.lua.org/) and demonstrates the flexibility of the architecture to support different program execution environments.
```
```suggestion
The system provides two _Runner_ implementations ([EVM](https://ethereum.org/en/developers/docs/evm/) and [Lua](https://www.lua.org/)) and demonstrates the flexibility of the architecture to support different program execution environments.
```

style nit that wasn't corrected.

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,165 @@
+
+Make sure you have [NodeJS installed](https://github.com/nodesource/distributions/blob/master/README.md).
+We reccomend using `v18` or `v20`, as Hardhat [actively supports these versions](https://github.com/nodejs/release#release-schedule) currently.
+We also provide a [.nvmrc](https://github.com/mit-dci/opencbdc-tx/blob/trunk/.nvmrc).
```
Thank you!

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,165 @@
+| New RPC URL     | `http://<agent IP address>:<agent port>/` |
+| Currency symbol | `CBDC`                                    |
+
+The agent IP address and agent port are the IP address and port specified
+for the agent when spawning the system.  
+Using Docker, this defaults to {IP: localhost, port: 8080}  
+Running outside of Docker, this defaults to {IP: localhost, port:8888}.
```
```suggestion
The agent IP address and agent port are the IP address and port specified for the agent when spawning the system.  
Using Docker, this defaults to {IP: localhost, port: 8080}.
Running outside of Docker, this defaults to {IP: localhost, port:8888}.
```
Another really odd line-breaking pattern.

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,165 @@
+Create a deploy script for the token. An example is provided in [opencbdc-tx/scripts/deploy.js](https://github.com/mit-dci/opencbdc-tx/blob/trunk/docs/programmability_user_guide.md).  
+As written, this script requires the name of the ERC20 token to be changed to match the name of the token which is to be deployed.  
+
+Note: When the Hardhat project was initialized, a `scripts/deploy.js` file was created,
+which corresponds to the file `contracts/Lock.sol`   
+that was also created upon initialization.
```
```suggestion
Note: When the Hardhat project was initialized, a `scripts/deploy.js` file was created.
It corresponds to the file `contracts/Lock.sol` that was also created upon initialization.
```

HalosGhost commented on docs/programmability_user_guide.md:
```diff
@@ -0,0 +1,165 @@
+## Running the Server
+### Run with Docker
+
+Spawns an agent running the EVM runner by default. Lua runner may be used by modifying the --runner_type
+flag in the command keyed by agent0, in the file [`docker-compose-3pc.yml`](https://github.com/mit-dci/opencbdc-tx/blob/trunk/docker-compose-3pc.yml).
```
```suggestion
Spawns an agent running the EVM runner by default.
Lua runner may be used by modifying the `--runner_type` flag in the command keyed by agent0, in the file [`docker-compose-3pc.yml`](https://github.com/mit-dci/opencbdc-tx/blob/trunk/docker-compose-3pc.yml).
```

HalosGhost commented on scripts/3pc-run-local.sh:
```diff
@@ -0,0 +1,48 @@
+echo Log level = $LOGLEVEL
+echo Runner type = $RUNNER_TYPE
+
+./build/src/3pc/runtime_locking_shard/runtime_locking_shardd --shard_count=1 --shard0_count=1 --shard00_endpoint=$IP:5556 --shard00_raft_endpoint=$IP:5557 --node_id=0 --component_id=0 --agent_count=1 --agent0_endpoint=$IP:6666 --ticket_machine_count=1 --ticket_machine0_endpoint=$IP:7777 --loglevel=$LOGLEVEL > logs/shardd.log &
+sleep 1
+./scripts/wait-for-it.sh -s $IP:5556 -t 60 -- ./build/src/3pc/ticket_machine/ticket_machined --shard_count=1 --shard0_count=1 --shard00_endpoint=$IP:5556 --node_id=0 --component_id=0 --agent_count=1 --agent0_endpoint=$IP:6666 --ticket_machine_count=1 --ticket_machine0_endpoint=$IP:7777 --loglevel=$LOGLEVEL > logs/ticket_machined.log &
+sleep 1
+./scripts/wait-for-it.sh -s $IP:7777 -t 60 -- ./scripts/wait-for-it.sh -s $IP:5556 -t 60 -- ./build/src/3pc/agent/agentd --shard_count=1 --shard0_count=1 --shard00_endpoint=$IP:5556 --node_id=0 --component_id=0 --agent_count=1 --agent0_endpoint=$IP:$PORT --ticket_machine_count=1 --ticket_machine0_endpoint=$IP:7777 --loglevel=$LOGLEVEL --runner_type=$RUNNER_TYPE > logs/agentd.log &
```
maintainability note (consider this slightly above a nit), it's dramatically harder to read really long source lines (which is why our linter forces line-breaks in the C++ code). For shell, this is traditionally improved by adding an escaped line-break after each option.

HalosGhost reviewed (approved):
I have a few very small style/typo nits, none of which are particularly high-priority.

Assuming the final bits of documentation follow all the same criteria current docs have gone through, that the renaming goes without a hitch (@maurermi and I chatted about strategy for this a bit offline, and I think there's a good plan in-place), and that the commits are squashed appropriately, I think this is set to merge. :+1:

maurermi committed:
Rename 3PC -> PArSEC

Rename all instances of 3PC to PArSEC.

Signed-off-by: Michael Maurer <maurer.mi@northeastern.edu>
Co-authored-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>
Co-authored-by: Anders Brownworth <anders-git@evantide.com>

maurermi reviewed (approved):
T-ACK. I think this is set to merge pending the name change.

AlexRamRam committed:
Fix in contract deployment

In evm_host::create(), the evmc::result return object must have the
create_address member set to the new contract address (the VM does
not set it)

If left as 0x0, deployment of contracts by another contract (i.e. at
depth>0) is likely to fail such as in UniswapV3Factory.createPool()

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

HalosGhost committed:
fix: reconcile docs for rename and updates

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

HalosGhost commented:
Finalized reconciliation of rename, and updated docs. :+1: from me. Ready for merge.
